lib/Yield/PTML.cpp

```diff
@@ -239,43 +241,97 @@ class InstructionPrefixManager {
            + std::move(Result);
   }
 
-  /// \note This does _not_ consume anything, feel free to call as many times
-  ///       as you need.
-  std::string emitEmpty(const ptml::MarkupBuilder &B,
-                        const model::Binary &Binary) {
-    uint64_t TotalPrefixSize = 2;
+  uint64_t totalPrefixSize(const model::Binary &Binary) const {
+    uint64_t Result = 2;
 
     if (LongestAddressString != 0) {
       using model::Architecture::getAssemblyLabelIndicator;
       auto Indicator = getAssemblyLabelIndicator(Binary.Architecture());
-      TotalPrefixSize += LongestAddressString + Indicator.size() + 4;
+      Result += LongestAddressString + Indicator.size() + 4;
     }
 
     if (LongestByteString != 0)
-      TotalPrefixSize += LongestByteString + 3;
+      Result += LongestByteString + 3;
+
+    return Result;
+  }
 
-    return B.getTag(tags::Span, std::string(TotalPrefixSize, ' '))
+  /// \note This does _not_ consume anything, feel free to call as many times
+  ///       as you need.
+  std::string emitEmpty(const ptml::MarkupBuilder &B,
+                        const model::Binary &Binary) const {
+    return B.getTag(tags::Span, std::string(totalPrefixSize(Binary), ' '))
       .addAttribute(attributes::Token, ptml::tokens::Indentation)
       .toString();
   }
 };
 
+struct StatementGraphNode {
+  const yield::BasicBlock *Block;
+
+  StatementGraphNode(const yield::BasicBlock &BB, const MetaAddress &) :
+    Block(&BB) {}
+};
+using StatementGraph = GenericGraph<ForwardNode<StatementGraphNode>>;
+
+template<>
+struct yield::StatementTraits<StatementGraph::Node *> {
+
+  using StatementType = const yield::Instruction *;
+
+  // TODO: switch to `std::optional` after updating to c++26.
+  using LocationType = const llvm::SmallVector<MetaAddress, 1>;
+
+  static auto getStatements(const StatementGraph::Node *Node) {
+    constexpr static SortedVector<yield::Instruction> Empty{};
+    return (Node->Block->ID().isValid() ? Node->Block->Instructions() : Empty)
+           | std::views::transform([](auto &&R) { return &R; });
+  }
```

Comment:
1. I've no idea what this method returns at this point.
   Can you use the explicit type instead of `auto`?
2. I suspect `->Instructions()` will be empty if `not ->ID().isValid()`.

Reply:
1. That's defined in the trait:
```cpp
  // static /* range_with_value_type<StatementType> auto */
  // getStatements(NodeType);
  {
    StatementTraits<NodeType>::getStatements(Node)
  } -> range_with_value_type<typename StatementTraits<NodeType>::StatementType>;
```

2. Yeah, but better be safe than sorry. I'd rather we didn't assume anything about a basic block without a valid address.

Reply:
1. That's far from being clear. What is the actual return type of `getStatements` in that specific trait specialization? I think it's a range of pointer to something. Let's try to be explicit, if the types are not too ugly.
2. If you want to be safe `assert`, do not try to handle "weird" cases.

Reply:
1.1. The thing is that we don't really care what the type is, as long as it can be iterated upon.
1.2. It is kind of ugly, `std::ranges::transform_view</* something something*/>`. I'm not sure how nice it is to rely on the `range` internals though.
2. 👌, assert works, I think.

Reply:
1. I agree that we don't care about what's the container we iterate on, but I don't know if what I'm iterating on is a pointer, a reference or what. That's confusing.
   Can you report the exact type? You can put `int` there to force the compiler giving you the exact type.

Reply:
Maybe I'm overdoing this, but I think it's worth getting a proper solution given the increased usage of ranges in our codebase.

Can we make it return `RangeOf<Type *> auto` ([example](https://godbolt.org/z/6d4jzrdT5))?
We erase the details of the actual type, but we still convey to the user what the function returns without looking at the function body.

Reply:
Ah, wait, you can do that? For some reason I was assuming it causes problems on return values. Yeah, my bad! Sorry you had to spell it out for me. This is definitely a huge improvement!

Reply:
It wasn't obvious for me either this was doable, but I tried and it just worked `:D`

Reply:
Done!